src/encodings/encoding.rs

```diff
@@ -331,6 +335,282 @@ default impl<T: DataType> Encoder<T> for DictEncoder<T> {
   }
 }
 
+// ----------------------------------------------------------------------
+// DELTA_BINARY_PACKED encoding
+
+const MAX_PAGE_HEADER_WRITER_SIZE: usize = 32;
+const MAX_BIT_WRITER_SIZE: usize = 10 * 1024 * 1024;
+const DEFAULT_BLOCK_SIZE: usize = 128;
+const DEFAULT_NUM_MINI_BLOCKS: usize = 4;
+
+/// Delta-binary-packing:
+///   [page-header] [block 1], [block 2], ... [block N]
+/// Each page header consists of:
+///   [block size] [number of miniblocks in a block] [total value count] [first value]
+/// Each block consists of:
+///   [min delta] [list of bitwidths of miniblocks] [miniblocks]
+///
+/// Current implementation writes values in 'put' method, multiple calls to 'put' add to existing
+/// block or start new block if block size is exceeded. Calling 'flush_buffer' writes out all data
+/// and resets internal state, including page header.
+///
+/// Supports only Int32Type and Int64Type.
+///
+pub struct DeltaBitPackEncoder<T: DataType> {
+  page_header_writer: BitWriter,
+  bit_writer: BitWriter,
+  total_values: usize,
+  first_value: i64,
+  current_value: i64,
+  block_size: usize,
+  mini_block_size: usize,
+  num_mini_blocks: usize,
+  values_in_block: usize,
+  deltas: Vec<i64>,
+  _phantom: PhantomData<T>
+}
+
+impl<T: DataType> DeltaBitPackEncoder<T> {
+  pub fn new() -> Self {
+    let block_size = DEFAULT_BLOCK_SIZE;
+    let num_mini_blocks = DEFAULT_NUM_MINI_BLOCKS;
+    let mini_block_size = block_size / num_mini_blocks;
+    assert!(mini_block_size % 8 == 0);
+    Self::assert_supported_type().unwrap();
+
+    DeltaBitPackEncoder {
+      page_header_writer: BitWriter::new(MAX_PAGE_HEADER_WRITER_SIZE),
+      bit_writer: BitWriter::new(MAX_BIT_WRITER_SIZE),
+      total_values: 0,
+      first_value: 0,
+      current_value: 0, // current value to keep adding deltas
+      block_size: block_size, // can write fewer values than block size for last block
+      mini_block_size: mini_block_size,
+      num_mini_blocks: num_mini_blocks,
+      values_in_block: 0, // will be at most block_size
+      deltas: vec![0; block_size],
+      _phantom: PhantomData
+    }
+  }
+
+  // write page header for blocks, this method is invoked when we are done encoding values
+  // it is also okay to encode when no values have been provided
+  fn write_page_header(&mut self) {
+    // write the size of each block
+    self.page_header_writer.put_vlq_int(self.block_size as u64);
+    // write the number of mini blocks
+    self.page_header_writer.put_vlq_int(self.num_mini_blocks as u64);
+    // write the number of all values (including non-encoded first value)
+    self.page_header_writer.put_vlq_int(self.total_values as u64);
+    // write first value
+    self.page_header_writer.put_zigzag_vlq_int(self.first_value);
```

Comment:
wonder if we should silently ignore the return values. Maybe just check the last one?

Reply:
Do you mean checking boolean result whether or not value was inserted and panic if it was not? If so, then yes, I can add the check and return Err() if `write_page_header()` method fails (fails to write at least one of those values).

Should I add similar check for all `bit_writer.put_***` methods?

Reply:
I agree, we should update - will fix.